app/src/test/java/illyan/jay/view_model/LoginViewModelTest.kt

```diff
@@ -0,0 +1,205 @@
+        every { mockPresenter.isUserLoggedIn() } returns false
+        every { mockPresenter.addAuthStateListener(any()) } answers {
+            firstArg<(Boolean) -> Unit>().invoke(true)
+        }
+
+        viewModel.observeStateAndEvents { stateObserver, _ ->
+            viewModel.load()
+            stateObserver.assertObserved(Initial, LoggedOut, LoggedIn)
+        }
+
+        verify(exactly = 1) { mockPresenter.isUserLoggedIn() }
+        verify(exactly = 1) { mockPresenter.addAuthStateListener(any()) }
+    }
+
+    @Test
+    fun load_changes_state_to_logged_out_when_user_logs_out() {
+        every { mockPresenter.isUserLoggedIn() } returns true
+        every { mockPresenter.addAuthStateListener(any()) } answers {
+            firstArg<(Boolean) -> Unit>().invoke(false)
+        }
+
+        viewModel.observeStateAndEvents { stateObserver, _ ->
+            viewModel.load()
+            stateObserver.assertObserved(Initial, LoggedIn, LoggedOut)
+        }
+
+        verify(exactly = 1) { mockPresenter.isUserLoggedIn() }
+        verify(exactly = 1) { mockPresenter.addAuthStateListener(any()) }
+    }
+
+    @Test
+    fun on_try_login_sets_state_to_logging_in() {
+        viewModel.observeStateAndEvents { stateObserver, _ ->
+            viewModel.onTryLogin()
+            stateObserver.assertObserved(Initial, LoggingIn)
+        }
+    }
+
+    @Test
+    fun google_sign_in_succeeds_when_auth_is_successful() = runTest {
+        val mockActivity: Activity = mockk()
+
+        val mockAccount: GoogleSignInAccount = mockk()
+        every { mockAccount.idToken } returns "MOCK_TOKEN"
+
+        val mockSignInTask: Task<GoogleSignInAccount> = mockk()
+        every { mockSignInTask.getResult(ApiException::class.java) } returns mockAccount
+
+        val mockCredential: AuthCredential = mockk()
+
+        val mockAuthResultTask: Task<AuthResult> = mockk()
+        every { mockAuthResultTask.addOnCompleteListener(mockActivity, any() ) } answers {
+            secondArg<OnCompleteListener<AuthResult>>().onComplete(mockAuthResultTask)
+            mockAuthResultTask
+        }
+        every { mockAuthResultTask.isSuccessful } returns true
+
+        mockkStatic(GoogleAuthProvider::class)
+        every { GoogleAuthProvider.getCredential("MOCK_TOKEN", null) } returns mockCredential
+
+        mockkStatic(FirebaseAuth::class)
+        every { FirebaseAuth.getInstance() } returns mockk(relaxed = true)
+        every { Firebase.auth.signInWithCredential(mockCredential) } returns mockAuthResultTask
+
+        viewModel.handleGoogleSignInResult(mockActivity, mockSignInTask)
+
+        verify(exactly = 1) { GoogleAuthProvider.getCredential("MOCK_TOKEN", null) }
+        verify(exactly = 1) { Firebase.auth.signInWithCredential(mockCredential) }
+        verify(exactly = 1) { mockAuthResultTask.addOnCompleteListener(mockActivity, any()) }
+        verify(exactly = 1) { mockTree.i("signInWithCredential:success") }
```

Comment:
Ilyen teszteknél lehetne használni a verifySequence vagy verifyOrder MockK függvényeket is, hogy a hívások sorrendje is tesztelésre kerüljön.

Reply:
Javítva.